infrastructure/src/test/java/org/corfudb/infrastructure/CorfuSequencerServerTest.java

```diff
@@ -0,0 +1,528 @@
+            Throwable payloadThrowable = (Throwable) ois.readObject();
+            assertTrue(payloadThrowable instanceof IllegalArgumentException);
+        } catch (Exception ex) {
+            fail("Missing an IllegalArgumentException from the SequencerServer");
+        }
+    }
+
+    /**
+     * Test that when client sends the SequencerMetricsRequestMsg, the server responds with
+     * SequencerMetricsResponseMsg with SequencerStatus.READY as its content.
+     */
+    @Test
+    public void sequencerMetricsRequest() {
+        sequencerServer = new SequencerServer(mockServerContext, spyFactoryHelper);
+        sequencerServer.setSequencerEpoch(0L);
+        RequestMsg request = getRequestMsg(
+                getBasicHeader(true, true),
+                CorfuProtocolSequencer.getSequencerMetricsRequestMsg()
+        );
+
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        sequencerServer.handleMessage(request, mockChannelHandlerContext, mockServerRouter);
+
+        verify(mockServerRouter).sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+        ResponseMsg response = responseCaptor.getValue();
+
+        // Assert that the payload has a getSequencerMetricsResponseMsg and that the base
+        // header fields have remained the same
+        assertTrue(compareBaseHeaderFields(request.getHeader(), response.getHeader()));
+        assertTrue(response.getPayload().hasSequencerMetricsResponse());
+
+        // Assert that the SequencerMetricsResponseMsg contains SequencerStatus.READY
+        SequencerMetricsResponseMsg sequencerMetricsResponseMsg = response.getPayload().getSequencerMetricsResponse();
+        assertTrue(sequencerMetricsResponseMsg.hasSequencerMetrics());
+        assertThat(sequencerMetricsResponseMsg.getSequencerMetrics().getSequencerStatus())
+                .isEqualTo(SequencerStatus.READY);
+    }
+
+    /**
+     * Test that when client sends the SequencerMetricsRequestMsg, the server responds with
+     * SequencerMetricsResponseMsg with SequencerStatus.READY as its content.
+     */
+    @Test
+    public void sequencerTrimRequest() {
+        // Bootstrap the Sequencer Server with a stream
+        UUID streamA = UUID.nameUUIDFromBytes("streamA".getBytes());
+        // Construct new tailMaps and create a to BootstrapSequencerRequestMsg
+        Map<UUID, StreamAddressSpace> tailMap = new HashMap<>();
+        long globalTail = 1;
+        // StreamAddressSpace streamAddressSpace = new StreamAddressSpace(Address.NON_ADDRESS,
+        //         Roaring64NavigableMap.bitmapOf(globalTail));
+        StreamAddressSpace streamAddressSpace = mock(StreamAddressSpace.class);
+        tailMap.put(streamA, streamAddressSpace);
+        long bootstrapMsgEpoch = 1;
+
+        SequencerServerCache cache = mock(SequencerServerCache.class);
+        SequencerServer.FactoryHelper mockFactoryHelper = mock(SequencerServer.FactoryHelper.class);
```

Comment:
Why do you have multiple factories? You have already used the spy annotation above.

Reply:
Good catch.
`when(mockSequencerFactoryHelper.getSequencerServerCache(anyInt(), anyLong())).thenReturn(cache);`
This line was failing with `IllegalArgumentException`  from `SequencerServerCache` class (or above levels) with spy one as it was actually invoking the method. Hence, I had to convert it into mock. I didn't want to change any of that code. I even tried some default values(ex - `DEFAULT_CACHE_SIZE = 250_000`) but to no avail.

Reply:
I'm not sure I understand where the exception comes from. Something like `doReturn(cache).when(mockSequencerFactoryHelper).getSequencerServerCache(anyInt(), anyLong())` should work. Alternatively, maybe we can always use a mock of the factory instead of a spy?

Note: https://stackoverflow.com/questions/11620103/mockito-trying-to-spy-on-method-is-calling-the-original-method provides a few possible explanations for the issue.